### PR TITLE
#12253: Batch Norm support for training mode

### LINKED ISSUE #16186: Running statistics updates in batch normalization
Provide support for updation of running stats in training mode for batch normalisation
`updated_running_stat = (1 − momentum) × running_stat + momentum × batch_stat`

### PR DESCRIPTION
**Pull request by VirdhatchaniKN**
### List of Github issues

- https://github.com/tenstorrent/tt-metal/issues/12253 
- https://github.com/tenstorrent/tt-metal/issues/15449
- https://github.com/tenstorrent/tt-metal/issues/16186

### Problem description
To implement batch normalization as device op

### What's changed
Work done : 

- Provided support for Training mode

### Checklist
All test have passed
- [x] [All post-commit tests](https://github.com/tenstorrent/tt-metal/actions/runs/12929324369)
- [x] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/runs/12929326611)
- [x] **(For models and ops writers)** Full [new models](https://github.com/tenstorrent/tt-metal/actions/workflows/full-new-models-suite.yaml) -  [Link to test](https://github.com/tenstorrent/tt-metal/actions/runs/12929328486)
- [x] [(Single-card) Demo tests](https://github.com/tenstorrent/tt-metal/actions/runs/12929329915)
- [x] [(Single-card) Device perf regressions](https://github.com/tenstorrent/tt-metal/actions/runs/12929331234)
- [x] [Single-card Model perf tests](https://github.com/tenstorrent/tt-metal/actions/runs/12929332745)

### CODE REVIEW

**Review by Aswinmcw — APPROVED** [review]
LGTM

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_device_operation.hpp [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        const Tensor& batch_mean;
+        const Tensor& batch_var;
+        std::optional<Tensor> running_mean;
+        std::optional<Tensor> running_var;
+    };
+
+    using spec_return_value_t = TensorSpec;
+    using tensor_return_value_t = Tensor;
+
+    struct RunningStatisticsFactory {
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_device_operation.hpp:32]: _ProgramFactory_ is the convention we are following if there is only one variant
if there is more than one then _specificcase_ProgramFactory_

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_running_statistics.cpp [hunk 2]
```diff
@@ -0,0 +1,136 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/dataflow/writer_running_statistics.cpp:1]: new copyright for 2025 in all newly added files.

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp [hunk 3]
```diff
@@ -5,23 +5,48 @@
...
-    // TODO: Implementation for training mode is in progress
-    TT_FATAL((!training), "Support currently provided for inference mode only");
+    if (training) {
+        Tensor batch_mean = mean_NHW(input, memory_config);
+        Tensor mean_sq = mean_NHW(ttnn::square(input, memory_config), memory_config);
+        Tensor batch_var =
+            ttnn::subtract(mean_sq, ttnn::square(batch_mean, memory_config), std::nullopt, memory_config);
+        Tensor stats =
+            ttnn::prim::running_statistics(batch_mean, batch_var, momentum, running_mean, running_var, memory_config);
+        return ttnn::prim::batch_norm(input, batch_mean, batch_var, eps, weight, bias, output, memory_config);
```
Comment (umadevimcw) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp:48]: Can we have at as single call of batch_norm in both training and inference?

**Review by umadevimcw — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_kernel.cpp [hunk 4]
```diff
@@ -0,0 +1,60 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <cstdint>
+#include "compute_kernel_api/eltwise_binary.h"
+#include "compute_kernel_api/tile_move_copy.h"
+#include "dprint.h"
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_kernel.cpp:8]: comment dprint ?

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_kernel.cpp [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            mul_tiles_to_cb(cb_tmp1, cb_old_running_mean, cb_tmp3, 0, 0, 1, 1);      // cb_tmp1 * running stats
+            add_tiles_to_cb(cb_tmp2, cb_tmp3, cb_updated_running_mean, 0, 0, 1, 1);  // cb_tmp2 * cb_tmp3
+        }
+        if constexpr (old_running_var_has_value) {
+            sub_tiles_to_cb(cb_one, cb_momentum, cb_tmp1, 0, 0, 0, 0);              // 1 - momentum
+            mul_tiles_to_cb(cb_momentum, cb_batch_var, cb_tmp2, 0, 0, 0, 1);        // momentum * batch stat
+            mul_tiles_to_cb(cb_tmp1, cb_old_running_var, cb_tmp3, 0, 0, 1, 1);      // cb_tmp1 * running stats
+            add_tiles_to_cb(cb_tmp2, cb_tmp3, cb_updated_running_var, 0, 0, 1, 1);  // cb_tmp2 * cb_tmp3
+        }
+        cb_pop_front(cb_one, 1);
```
Comment (KalaivaniMCW) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/kernels/compute/running_statistics_kernel.cpp:51]: pop_front can be done after  tile_regs_release(); ?

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp [hunk 6]
```diff
@@ -1,31 +1,56 @@
...
+    auto batch_mean = input_tensor;
+    ttnn::SmallVector<int64_t> dims = {0, 2, 3};
+    std::sort(dims.begin(), dims.end());
+    for (uint32_t i = dims.size() - 1; i > 0; i--) {
+        auto temp_output = ttnn::prim::moreh_mean(
+            batch_mean, dims[i], true, std::nullopt, std::nullopt, output_memory_config, std::nullopt);
+        batch_mean = temp_output;
+    }
+    return ttnn::prim::moreh_mean(
+        batch_mean, dims.front(), true, std::nullopt, std::nullopt, output_memory_config, std::nullopt);
```
Comment (patrickroberts) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp:27]: This dependency concerns me, can we make sure moving forward that we don't depend on Moreh ops in our implementation? It creates the possibility of cyclic dependencies in the build which often become a headache to deal with.

I think either we should create an issue to eventually address this and add a `// TODO: remove dependency on Moreh` here, or figure out how to resolve it now.

@bbradelTT any thoughts on this?

Comment (bbradelTT) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp:27]: We should avoid the dependence on moreh where we can.

@VirdhatchaniKN Why can't we use ttnn::mean?

Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp:27]: I can modify the code accordingly

Comment (VirdhatchaniKN) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/batch_norm.cpp]: @bbradelTT  @patrickroberts 
I have updated the code with ttnn::mean. Would like your review here

**Review by patrickroberts — APPROVED** [review]

**Review by mcw-anasuya — APPROVED** [review]
LGTM

**Review by bbradelTT — APPROVED** [review]

**Review by yugaoTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp [hunk 7]
```diff
@@ -0,0 +1,318 @@
...
+#include <tt-metalium/work_split.hpp>
+#include "ttnn/operations/cb_utils.hpp"
+#include <cmath>
+
+namespace {
+namespace CMAKE_UNIQUE_NAMESPACE {
+
+using namespace ttnn::operations::normalization;
+
+std::tuple<uint32_t, uint32_t, uint32_t, uint32_t> extract_shape_dims(const Tensor& x) {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp:16]: @sminakov-tt  ptal

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+#include "running_statistics_device_operation.hpp"
+
+#include <tt-metalium/work_split.hpp>
+#include "ttnn/operations/cb_utils.hpp"
+#include <cmath>
+
+namespace {
+namespace CMAKE_UNIQUE_NAMESPACE {
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp:12]: why do you want to use it?

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp [hunk 9]
```diff
@@ -0,0 +1,318 @@
...
+namespace ttnn::operations::normalization {
+RunningStatistics::RunningStatisticsProgramFactory::cached_program_t
+RunningStatistics::RunningStatisticsProgramFactory::create(
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& output) {
+    using namespace tt;
+    using namespace tt::tt_metal;
+
+    const auto& [a, b, d, e] = tensor_args;
```
Comment (dmakoviichuk-tt) [inline ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp:131]: is there a way to get a better names for tensors then a,b,d,e? Also why c is missed?

File: ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp (hunk 8 above)
Comment (mouliraj-mcw) [reply ttnn/cpp/ttnn/operations/normalization/batch_norm/device/running_statistics_program_factory.cpp:12]: It is a namespace set in Unity's CMake configuration and is included across all program factories.

**Review by mouliraj-mcw — APPROVED** [review]

**Review by dmakoviichuk-tt — APPROVED** [review]

**Comment by VirdhatchaniKN** [comment]
Passing CI Tests : 

- https://github.com/tenstorrent/tt-metal/actions/runs/12995222930
- https://github.com/tenstorrent/tt-metal/actions/runs/12995226539
- https://github.com/tenstorrent/tt-metal/actions/runs/12995228264
- https://github.com/tenstorrent/tt-metal/actions/runs/12995230126

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/13031507387 - test passed as in main

**Comment by VirdhatchaniKN** [comment]
Test file passes. will merge after PR Checks have completed. waiting for https://github.com/tenstorrent/tt-metal/actions/runs/13044600868/job/36392961266?pr=16592

**Comment by VirdhatchaniKN** [comment]
https://github.com/tenstorrent/tt-metal/actions/runs/13098157794 - passed as in main
